File: interop_client.py
#!/usr/bin/env python3
"""
interop_client.py
==================
A standalone Python client speaking network_os.py's wire protocol
(PROTOCOL.md) -- written WITHOUT importing network_os.py or
crypto_layer.py. The point isn't to reuse the reference implementation's
own code path; it's to prove PROTOCOL.md alone is enough to build a
conforming client, in the same language as the reference implementation.

Uses the `cryptography` package (the same dependency network_os.py
itself uses) for Ed25519 / X25519 / HKDF-SHA256 / AES-256-GCM.

Run: python3 interop_client.py 127.0.0.1 9501
"""
from __future__ import annotations
import hashlib
import json
import os
import socket
import sys
import time

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey, Ed25519PublicKey,
)
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey, X25519PublicKey,
)
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def main() -> None:
    host = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 9501

    node_id = sha256_hex(b"py-interop-node")[:16]
    strand_hex = "0" * 64

    signing_priv = Ed25519PrivateKey.generate()
    signing_pub = signing_priv.public_key()
    exchange_priv = X25519PrivateKey.generate()
    exchange_pub = exchange_priv.public_key()

    my_signing_pub_hex = signing_pub.public_bytes(Encoding.Raw, PublicFormat.Raw).hex()
    my_exchange_pub_hex = exchange_pub.public_bytes(Encoding.Raw, PublicFormat.Raw).hex()

    to_sign = f"{node_id}|{strand_hex}|{my_exchange_pub_hex}".encode()
    sig = signing_priv.sign(to_sign)

    hello = {
        "type": "hello", "node_id": node_id, "listen_port": 0, "strand_hex": strand_hex,
        "signing_pub": my_signing_pub_hex, "exchange_pub": my_exchange_pub_hex, "sig": sig.hex(),
    }

    print(f"[py] connecting to {host}:{port}")
    sock = socket.create_connection((host, port))
    sock.sendall((json.dumps(hello) + "\n").encode())
    print(f"[py] sent hello (node_id={node_id})")

    buf = b""
    while b"\n" not in buf:
        chunk = sock.recv(65536)
        if not chunk:
            break
        buf += chunk
    line, _, _ = buf.partition(b"\n")
    print("[py] received:", line.decode())
    ack = json.loads(line)

    to_verify = f"{ack['node_id']}|{ack['strand_hex']}|{ack['exchange_pub']}".encode()
    peer_signing_pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(ack["signing_pub"]))
    try:
        peer_signing_pub.verify(bytes.fromhex(ack["sig"]), to_verify)
        sig_ok = True
    except Exception:
        sig_ok = False
    print("[py] Python node signature verified:", sig_ok)
    if not sig_ok:
        sock.close()
        sys.exit(1)

    peer_exchange_pub = X25519PublicKey.from_public_bytes(bytes.fromhex(ack["exchange_pub"]))
    shared_secret = exchange_priv.exchange(peer_exchange_pub)

    salt = b"\x00" * 32  # matches HKDF's salt=None default (HashLen zero bytes, RFC 5869)
    session_key = HKDF(
        algorithm=hashes.SHA256(), length=32, salt=salt, info=b"network-os-session-v1",
    ).derive(shared_secret)
    print("[py] derived session key:", session_key.hex())

    block_id = "py-000001"
    source = "py_interop_node"
    feature_hash = sha256_hex(b"py-originated-event")
    confidence = 1
    timestamp = int(time.time())

    canonical = (
        f'{{"block_id": "{block_id}", "confidence": {confidence}, '
        f'"feature_hash": "{feature_hash}", "source": "{source}", "timestamp": {timestamp}}}'
    )
    block_sig = signing_priv.sign(canonical.encode())

    signed_block = {
        "block_id": block_id, "confidence": confidence, "feature_hash": feature_hash,
        "source": source, "timestamp": timestamp, "sig": block_sig.hex(),
    }
    inner_plaintext = json.dumps({"type": "block", "block": signed_block})

    nonce = os.urandom(12)
    ciphertext = AESGCM(session_key).encrypt(nonce, inner_plaintext.encode(), node_id.encode())
    payload = nonce + ciphertext

    envelope = {"type": "enc", "node_id": node_id, "payload": payload.hex()}
    sock.sendall((json.dumps(envelope) + "\n").encode())
    print("[py] sent signed, AES-256-GCM-encrypted block to Python node")

    time.sleep(0.3)
    sock.close()
    print("[py] done, closing")


if __name__ == "__main__":
    main()

File: interop_client.js
// interop_client.js
// A real Node.js client speaking network_os.py's wire protocol (PROTOCOL.md).
// Uses Node's built-in crypto module (Ed25519, X25519, AES-256-GCM,
// SHA-256) and the net module for the TCP transport. No npm dependencies.
// Requires Node.js 13.9+ (crypto.diffieHellman on X25519 KeyObjects).
//
// Run: node interop_client.js 127.0.0.1 9501

'use strict';
const crypto = require('crypto');
const net = require('net');

function sha256Hex(data) {
  return crypto.createHash('sha256').update(data).digest('hex');
}

// HKDF-SHA256 (RFC 5869) -- matching crypto_layer.py's derive_shared_key
// exactly. Hand-rolled via createHmac rather than crypto.hkdfSync so this
// also runs on Node versions older than the built-in HKDF API (15+).
function hkdfSha256(ikm, salt, infoStr, length) {
  const info = Buffer.from(infoStr);
  const prk = crypto.createHmac('sha256', salt).update(ikm).digest();
  let t = Buffer.alloc(0);
  let okm = Buffer.alloc(0);
  const n = Math.ceil(length / 32);
  for (let i = 1; i <= n; i++) {
    t = crypto.createHmac('sha256', prk)
      .update(Buffer.concat([t, info, Buffer.from([i])]))
      .digest();
    okm = Buffer.concat([okm, t]);
  }
  return okm.subarray(0, length);
}

// raw 32-byte public keys travel hex-encoded on the wire; Node's KeyObject
// API only speaks X.509/SPKI DER, so wrap/unwrap with the standard SPKI
// prefix for each key type (same trick the Ruby/C++ clients use).
const ED25519_DER_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');
const X25519_DER_PREFIX = Buffer.from('302a300506032b656e032100', 'hex');

function rawPubHex(keyObject) {
  const der = keyObject.export({ type: 'spki', format: 'der' });
  return der.subarray(der.length - 32).toString('hex');
}

function ed25519PubFromRawHex(hex) {
  const der = Buffer.concat([ED25519_DER_PREFIX, Buffer.from(hex, 'hex')]);
  return crypto.createPublicKey({ key: der, format: 'der', type: 'spki' });
}

function x25519PubFromRawHex(hex) {
  const der = Buffer.concat([X25519_DER_PREFIX, Buffer.from(hex, 'hex')]);
  return crypto.createPublicKey({ key: der, format: 'der', type: 'spki' });
}

const host = process.argv[2] || '127.0.0.1';
const port = parseInt(process.argv[3] || '9501', 10);

const nodeId = sha256Hex('js-interop-node').slice(0, 16);
const strandHex = '0'.repeat(64);

const { publicKey: signingPub, privateKey: signingPriv } = crypto.generateKeyPairSync('ed25519');
const { publicKey: exchangePub, privateKey: exchangePriv } = crypto.generateKeyPairSync('x25519');

const mySigningPubHex = rawPubHex(signingPub);
const myExchangePubHex = rawPubHex(exchangePub);

const toSign = `${nodeId}|${strandHex}|${myExchangePubHex}`;
const sig = crypto.sign(null, Buffer.from(toSign), signingPriv);

const hello = {
  type: 'hello', node_id: nodeId, listen_port: 0, strand_hex: strandHex,
  signing_pub: mySigningPubHex, exchange_pub: myExchangePubHex, sig: sig.toString('hex'),
};

console.log(`[js] connecting to ${host}:${port}`);
const socket = net.createConnection({ host, port }, () => {
  socket.write(JSON.stringify(hello) + '\n');
  console.log(`[js] sent hello (node_id=${nodeId})`);
});

let buffered = '';
socket.on('data', (chunk) => {
  buffered += chunk.toString('utf8');
  const nl = buffered.indexOf('\n');
  if (nl === -1) return;
  const line = buffered.slice(0, nl);
  buffered = buffered.slice(nl + 1);
  console.log('[js] received:', line);

  const ack = JSON.parse(line);
  const toVerify = `${ack.node_id}|${ack.strand_hex}|${ack.exchange_pub}`;
  const peerSigningPub = ed25519PubFromRawHex(ack.signing_pub);
  const sigOk = crypto.verify(null, Buffer.from(toVerify), peerSigningPub, Buffer.from(ack.sig, 'hex'));
  console.log('[js] Python node signature verified:', sigOk);
  if (!sigOk) { socket.destroy(); process.exit(1); }

  const peerExchangePub = x25519PubFromRawHex(ack.exchange_pub);
  const sharedSecret = crypto.diffieHellman({ privateKey: exchangePriv, publicKey: peerExchangePub });

  const salt = Buffer.alloc(32); // 32 zero bytes -- matches HKDF's salt=None default (RFC 5869)
  const sessionKey = hkdfSha256(sharedSecret, salt, 'network-os-session-v1', 32);
  console.log('[js] derived session key:', sessionKey.toString('hex'));

  const blockId = 'js-000001';
  const source = 'js_interop_node';
  const featureHash = sha256Hex('js-originated-event');
  const confidence = 1;
  const timestamp = Math.floor(Date.now() / 1000);

  const canonical = `{"block_id": "${blockId}", "confidence": ${confidence}, "feature_hash": "${featureHash}", "source": "${source}", "timestamp": ${timestamp}}`;
  const blockSig = crypto.sign(null, Buffer.from(canonical), signingPriv);

  const signedBlock = {
    block_id: blockId, confidence, feature_hash: featureHash,
    source, timestamp, sig: blockSig.toString('hex'),
  };
  const innerPlaintext = JSON.stringify({ type: 'block', block: signedBlock });

  const nonce = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', sessionKey, nonce);
  cipher.setAAD(Buffer.from(nodeId));
  const ciphertext = Buffer.concat([cipher.update(innerPlaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();

  const payload = Buffer.concat([nonce, ciphertext, tag]);
  const envelope = { type: 'enc', node_id: nodeId, payload: payload.toString('hex') };
  socket.write(JSON.stringify(envelope) + '\n');
  console.log('[js] sent signed, AES-256-GCM-encrypted block to Python node');

  setTimeout(() => {
    socket.end();
    console.log('[js] done, closing');
  }, 300);
});

socket.on('error', (err) => {
  console.error('[js] connect failed:', err.message);
  process.exit(1);
});
